DREAM/Optimization/tsgGradientDescent.cpp

```diff
@@ -0,0 +1,126 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * UT-BATTELLE, LLC AND THE UNITED STATES GOVERNMENT MAKE NO REPRESENTATIONS AND DISCLAIM ALL WARRANTIES, BOTH EXPRESSED AND
+ * IMPLIED. THERE ARE NO EXPRESS OR IMPLIED WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE, OR THAT THE USE OF
+ * THE SOFTWARE WILL NOT INFRINGE ANY PATENT, COPYRIGHT, TRADEMARK, OR OTHER PROPRIETARY RIGHTS, OR THAT THE SOFTWARE WILL
+ * ACCOMPLISH THE INTENDED RESULTS OR THAT THE SOFTWARE OR ITS USE WILL NOT RESULT IN INJURY OR DAMAGE. THE USER ASSUMES
+ * RESPONSIBILITY FOR ALL LIABILITIES, PENALTIES, FINES, CLAIMS, CAUSES OF ACTION, AND COSTS AND EXPENSES, CAUSED BY, RESULTING
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+#ifndef __TASMANIAN_GRADIENT_DESCENT_CPP
+#define __TASMANIAN_GRADIENT_DESCENT_CPP
+
+#include "tsgGradientDescent.hpp"
+
+namespace TasOptimization {
+
+GradientDescentState::GradientDescentState(const std::vector<double> &x0, const double lambda0) :
+        num_dimensions((int) x0.size()), adaptive_stepsize(lambda0), x(x0) {};
+
+OptimizationStatus GradientDescent(const ObjectiveFunctionSingle &f, const GradientFunctionSingle &g, const ProjectionFunctionSingle &proj,
+                                   GradientDescentState &state, const std::vector<double> &line_search_coeffs, const int num_iterations,
+                                   const double tolerance) {
+
+    if (line_search_coeffs.size() != 2)
+        throw std::runtime_error("ERROR: in GradientDescent(), expects line_search_coeffs.size() == 2");
+
+    OptimizationStatus status;
+    status.num_iterations = 0;
+    status.stationarity_residual = std::numeric_limits<double>::max();
+
+    int num_dimensions = state.num_dimensions;
+    std::vector<double> &x = state.getXRef();
+    std::vector<double> x_prev(x), g_at_x_prev(g(x_prev)), g_at_x(num_dimensions), next_x(num_dimensions);
+    double f_at_x_prev(f(x_prev)), f_at_x(0);
+
+    while(status.num_iterations < num_iterations) {
+        // Find the next stepsize/candidate point by making sure it satisfies the well-known descent inequality (see γ_u in the
+        // reference paper associated with this function).
+        double lhs = 0;
+        double rhs = 0;
+        do {
+            if (status.num_iterations >= num_iterations)
+                break;
+            for (int j=0; j<num_dimensions; j++)
+                x[j] = x_prev[j] - g_at_x_prev[j] * state.adaptive_stepsize;
+            x = proj(x);
+            f_at_x = f(x);
+            lhs += f_at_x - f_at_x_prev;
+            for (int j=0; j<num_dimensions; j++) {
+                double delta = x[j] - x_prev[j];
+                rhs += delta * delta / (2.0 * state.adaptive_stepsize);
```

Comment:
This saves one flop, how much does this depend on the projection of `x`, can we do it here?
```
double ngrad = g_at_x_prev[j] * g_at_x_prev[j] * state.adaptive_stepsize;
rhs += 0.5 * ngrad;
lhs -= ngrad;
```